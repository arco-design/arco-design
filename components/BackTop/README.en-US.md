`````
Component / Other

# BackTop

`BackTop` makes it easy to go back to the top of the page.
`````

%%Content%%

## API

### BackTop

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|duration|Duration to scroll to the top|number |`400`|
|visibleHeight|When scrolling to this height, display the back to top button.|number |`400`|
|easing|Scroll to the top of the easing method type, all types: [easing](https://github.com/PengJiyuan/b-tween).|string |`quartOut`|
|className|Additional css class|string \| string[] |`-`|
|style|Additional style|CSSProperties |`-`|
|onClick|Callback when you click to return to the top.|() => void |`-`|
|target|Set the element whose scroll event needs to be listened to. The value is a function that returns the corresponding `DOM` element|() => HTMLElement \| Window |`() => window`|
