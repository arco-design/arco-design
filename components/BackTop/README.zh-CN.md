`````
组件 / 其他

# 返回顶部 BackTop

可一键返回顶部的按钮。
`````

%%Content%%

## API

### BackTop

|参数名|描述|类型|默认值|
|---|---|---|---|
|duration|滚动到顶部的时间。|number |`400`|
|visibleHeight|当滚动到这个高度时，显示返回顶部按钮。|number |`400`|
|easing|滚动到顶部的缓动方式类型, 所有类型：[easing](https://github.com/PengJiyuan/b-tween)。|string |`quartOut`|
|className|节点类名|string \| string[] |`-`|
|style|节点样式|CSSProperties |`-`|
|onClick|点击返回顶部时的回调函数。|() => void |`-`|
|target|设置需要监听其滚动事件的元素，值为一个返回对应 `DOM` 元素的函数。|() => HTMLElement \| Window |`() => window`|
