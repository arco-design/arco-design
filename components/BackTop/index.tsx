import React, { PropsWithChildren, forwardRef, useState, useEffect, useContext, memo } from 'react';
import { CSSTransition } from 'react-transition-group';
import BTween from 'b-tween';
import { pickDataAttributes } from '../_util/pick';
import cs from '../_util/classNames';
import IconToTop from '../../icon/react-icon/IconToTop';
import { ConfigContext } from '../ConfigProvider';
import { on, off } from '../_util/dom';
import throttleByRaf from '../_util/throttleByRaf';
import { BackTopProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';
import useKeyboardEvent from '../_util/hooks/useKeyboardEvent';

const defaultProps: BackTopProps = {
  visibleHeight: 400,
  easing: 'quartOut',
  duration: 400,
  target: () => window,
};

function BackTop(baseProps: PropsWithChildren<BackTopProps>, ref) {
  const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);
  const getKeyboardEvents = useKeyboardEvent();
  const props = useMergeProps<PropsWithChildren<BackTopProps>>(
    baseProps,
    defaultProps,
    componentConfig?.BackTop
  );

  const prefixCls = getPrefixCls('backtop');

  const [visible, setVisible] = useState(false);

  const getTarget = (target: HTMLElement | Window): HTMLElement => {
    return target === window ? document.documentElement : (target as HTMLElement);
  };

  useEffect(() => {
    const target = props.target && props.target();
    const scrollHandler = throttleByRaf(() => {
      const visibleHeight = props.visibleHeight;
      const scrollTop = getTarget(target).scrollTop;
      setVisible(scrollTop >= visibleHeight);
    });

    on(target, 'scroll', scrollHandler);

    scrollHandler();

    return () => {
      scrollHandler.cancel && scrollHandler.cancel();
      off(target, 'scroll', scrollHandler);
    };
  }, [props.target, props.visibleHeight]);

  const scrollToTop = () => {
    const targetDom = props.target && props.target();
    const t = getTarget(targetDom);
    const scrollTop = t.scrollTop;
    const tween = new BTween({
      from: { scrollTop },
      to: { scrollTop: 0 },
      easing: props.easing,
      duration: props.duration,
      onUpdate: (keys) => {
        t.scrollTop = keys.scrollTop;
      },
    });
    tween.start();
    props.onClick && props.onClick();
  };

  return (
    <div
      {...pickDataAttributes(props)}
      ref={ref}
      className={cs(`${prefixCls}`, { [`${prefixCls}-rtl`]: rtl }, props.className)}
      style={props.style}
      onClick={scrollToTop}
      {...getKeyboardEvents({
        onPressEnter: scrollToTop,
      })}
    >
      <CSSTransition in={visible} timeout={100} classNames="fadeIn" unmountOnExit>
        {props.children || (
          <button className={`${prefixCls}-button`}>
            <IconToTop />
          </button>
        )}
      </CSSTransition>
    </div>
  );
}

const BackTopRef = forwardRef<unknown, PropsWithChildren<BackTopProps>>(BackTop);

BackTopRef.displayName = 'BackTop';

export default memo(BackTopRef);

export { BackTopProps };
