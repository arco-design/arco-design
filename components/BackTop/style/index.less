@import './token.less';

@backtop-prefix-cls: ~'@{prefix}-backtop';

.@{backtop-prefix-cls} {
  position: fixed;
  bottom: @backtop-margin-bottom;
  right: @backtop-margin-right;
  z-index: 100;
  cursor: pointer;

  &-button {
    width: @backtop-button-size-width;
    height: @backtop-button-size-width;
    font-size: @backtop-button-size-font;
    text-align: center;
    outline: none;
    background-color: @backtop-button-color-bg;
    border-radius: @backtop-button-border-radius;
    color: @backtop-button-color-text;
    transition: all @transition-duration-2 @transition-timing-function-linear;
    cursor: pointer;
    border: none;

    &:focus-visible {
      box-shadow: 0 0 0 2px var(~'@{arco-cssvars-prefix}-color-primary-light-3');
    }

    &:hover {
      background-color: @backtop-button-color-bg_hover;
    }

    svg {
      font-size: @backtop-button-size-icon;
    }
  }
}
