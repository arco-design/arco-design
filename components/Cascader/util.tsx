import { isArray } from '../_util/is';
import Store from './base/store';
import { CascaderProps } from './interface';

export const ValueSeparator = '__arco_cascader__';

export const SHOW_PARENT = 'parent';
export const SHOW_CHILD = 'child';

export const PANEL_MODE = {
  cascader: 'cascader',
  select: 'select',
};

export function isEmptyValue(value) {
  return !value || (isArray(value) && value.length === 0);
}

export function getConfig(props: CascaderProps) {
  return {
    showEmptyChildren: props.showEmptyChildren,
    changeOnSelect: props.changeOnSelect,
    lazyload: !!props.loadMore,
    fieldNames: props.fieldNames,
    filterOption: props.filterOption,
    showParent:
      props.mode === 'multiple' && !props.changeOnSelect && props.checkedStrategy === SHOW_PARENT,
  };
}

export function getStore(props, value) {
  const tmp = value ? (Array.isArray(value[0]) ? value : [value]) : [];
  return new Store(props.options || [], tmp, getConfig(props));
}

export const transformValuesToSet = (values: string[][]) => {
  const _values = values || [];
  const valuesSet = _values.reduce((set, next) => {
    // 'next' could be a string.
    set.add([].concat(next).join(ValueSeparator));
    return set;
  }, new Set());

  return valuesSet;
};

export const valueInSet = (set, value: string[]) => {
  const _value = value || [];
  return set.has(_value.join(ValueSeparator));
};

export const removeValueFromSet = (set, value: string[]) => {
  const _value = value || [];
  return set.delete(_value.join(ValueSeparator));
};

export const formatValue = (value, isMultiple, store?): string[][] | undefined => {
  let _value = [];
  if (value === undefined) {
    _value = [];
  } else if (isMultiple) {
    _value = value;
  } else {
    _value = [value];
  }

  if (store && store.config.showParent) {
    const checkedNodes = store.getCheckedNodes();
    const valuesSet = transformValuesToSet(checkedNodes.map((node) => node.pathValue));
    const result = [];
    const temp = {};
    _value.map((v) => {
      v.some((_, index, arr) => {
        const curVal = arr.slice(0, index + 1);
        const pass = valueInSet(valuesSet, curVal);
        if (pass && !temp[curVal.join(ValueSeparator)]) {
          result.push(curVal);
          temp[curVal.join(ValueSeparator)] = 1;
        }
        return pass;
      });
    });

    return result;
  }
  return _value;
};

// change check status to false
const deny2Checked = (option) => {
  const deny = (options) => {
    return !Array.isArray(options)
      ? false
      : options.every((item) => {
          if (item._checked || item.disabled) {
            return true;
          }
          return deny(item.children);
        });
  };
  return option._halfChecked && deny(option?.children);
};

export const getMultipleCheckValue = (propsValue, store: Store<any>, option, _checked) => {
  const checked = _checked && deny2Checked(option) ? false : _checked;

  const beforeValueSet = store.getCheckedNodes().reduce((set, node) => {
    set.add(node.pathValue.join(ValueSeparator));
    return set;
  }, new Set());

  option.setCheckedState(checked);
  const checkedNodes = store.getCheckedNodes();
  const currentValue = checkedNodes.map((node) => node.pathValue);
  const currentValueSet = transformValuesToSet(currentValue);

  const newValueSet = new Set();
  return propsValue
    .filter((v) => {
      // v 不在 beforeValueSet 中，说明 v 不包含对应的option。直接返回true，不应该清除掉。
      if (!valueInSet(beforeValueSet, v) || valueInSet(currentValueSet, v)) {
        newValueSet.add(v.join(ValueSeparator));
        return true;
      }
    })
    .concat(
      currentValue.filter((v) => {
        return !valueInSet(newValueSet, v);
      })
    );
};
