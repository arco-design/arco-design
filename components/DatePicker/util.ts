import dayjs, { Dayjs } from 'dayjs';
import { isArray, isDayjs } from '../_util/is';
import { methods } from '../_util/dayjs';

function getFormat(time) {
  return isDayjs(time) && time.format('HH:mm:ss');
}

export function isTimeArrayChange(prevTime: Dayjs[], nextTime: Dayjs[]) {
  return (
    getFormat(prevTime[0]) !== getFormat(nextTime[0]) ||
    getFormat(prevTime[1]) !== getFormat(nextTime[1])
  );
}

export function getAvailableDayjsLength(value) {
  if (!value) {
    return 0;
  }
  if (isArray(value)) {
    if (isDayjs(value[0]) && isDayjs(value[1])) {
      return 2;
    }
    if (!isDayjs(value[0]) && !isDayjs(value[1])) {
      return 0;
    }
    return 1;
  }
  return 0;
}

// https://github.com/react-component/picker/blob/master/src/utils/dateUtil.ts#L234
export function isDisabledDate(cellDate, disabledDate, mode): boolean {
  if (typeof disabledDate !== 'function') {
    return false;
  }
  // Whether cellDate is disabled in range
  const getDisabledFromRange = (
    currentMode: 'date' | 'month' | 'year',
    start: number,
    end: number
  ) => {
    let current = start;
    while (current <= end) {
      let date: Dayjs;
      switch (currentMode) {
        case 'date': {
          date = methods.set(cellDate, 'date', current);
          if (!disabledDate(date)) {
            return false;
          }
          break;
        }
        case 'month': {
          date = methods.set(cellDate, 'month', current);
          if (!isDisabledDate(date, disabledDate, 'month')) {
            return false;
          }
          break;
        }
        case 'year': {
          date = methods.set(cellDate, 'year', current);
          if (!isDisabledDate(date, disabledDate, 'year')) {
            return false;
          }
          break;
        }
        default:
          break;
      }
      current += 1;
    }
    return true;
  };
  switch (mode) {
    case 'date':
    case 'week': {
      return disabledDate(cellDate);
    }
    case 'month': {
      const startDate = 1;
      const endDate = cellDate.endOf('month').get('date');
      return getDisabledFromRange('date', startDate, endDate);
    }
    case 'quarter': {
      const startMonth = Math.floor(cellDate.get('month') / 3) * 3;
      const endMonth = startMonth + 2;
      return getDisabledFromRange('month', startMonth, endMonth);
    }
    case 'year': {
      return getDisabledFromRange('month', 0, 11);
    }
    default:
      return false;
  }
}

type WeekStartType = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export function getDefaultWeekStart(dayjsLocale: string): WeekStartType {
  return (dayjs.Ls?.[dayjsLocale]?.weekStart as WeekStartType) || 0;
}

export function getLocaleDayjsValue(
  date: Dayjs | undefined,
  dayjsLocale: string
): Dayjs | undefined {
  return date ? date.locale(dayjsLocale) : date;
}
