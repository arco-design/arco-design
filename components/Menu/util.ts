import React, { ReactElement } from 'react';

export type MenuInfo = {
  keyPath: string[];
  prev: string;
  next: string;
  firstChild?: string;
  lastChild?: string;
  disabled?: boolean;
};

export const PROPS_NEED_TO_BE_PASSED_IN_SUBMENU = ['popup', 'triggerProps', 'selectable'];

// Expand MenuGroup to get an array only contains MenuItem and SubMenu
const flatMenuGroup = (children): ReactElement[] => {
  let validMenuItems = [];

  React.Children.forEach(children, (item) => {
    const menuType = item?.type?.menuType;
    if (menuType === 'MenuItem' || menuType === 'SubMenu') {
      validMenuItems.push(item);
    } else if (menuType === 'MenuGroup') {
      validMenuItems = validMenuItems.concat(flatMenuGroup(item.props.children));
    }
  });

  return validMenuItems;
};

export const generateInfoMap = (
  children,
  keyPath: string[] = [],
  result: {
    [key: string]: MenuInfo;
  } = {}
) => {
  const validChildrenList: any[] = flatMenuGroup(children);

  validChildrenList.forEach((item, index) => {
    const key = item.key;
    const menuType = item.type.menuType;
    const _keyPath = [key, ...keyPath];
    const info: MenuInfo = {
      keyPath: [],
      prev: validChildrenList[index - 1]?.key || null,
      next: validChildrenList[index + 1]?.key || null,
    };

    if (index === 0 || index === validChildrenList.length - 1) {
      const parentKey = _keyPath[1];
      const propertyName = index === 0 ? 'firstChild' : 'lastChild';
      if (parentKey) {
        result[parentKey] = {
          ...result[parentKey],
          [propertyName]: key,
        };
      }
    }

    switch (menuType) {
      case 'SubMenu':
        info.keyPath = _keyPath;
        generateInfoMap(item.props.children, _keyPath, result);
        break;

      case 'MenuItem':
        info.keyPath = _keyPath;
        info.disabled = item.props.disabled;
        break;

      default:
        break;
    }

    result[key] = {
      ...result[key],
      ...info,
    };
  });

  return result;
};

export const processChildren = (
  children,
  props
): Array<Exclude<unknown, boolean | null | undefined>> => {
  return React.Children.map(children, (item, index) => {
    if (!item || !item.props) {
      return item;
    }

    const isHTMLElement = typeof item.type === 'string';
    const isMenuSubComponent = item.type && item.type.menuType;

    // 处理 <div> 包裹 MenuItem 之类的情况
    if (!isMenuSubComponent && item.props.children) {
      const _props = isHTMLElement ? {} : props;
      return React.cloneElement(item, {
        ..._props,
        _key: item.key,
        children: processChildren(item.props.children, props),
      });
    }

    return isHTMLElement
      ? item
      : React.cloneElement(item, {
          ...props,
          // Properties of the component itself have higher priority
          ...item.props,
          _key: item.key || `$menu-${index}`,
        });
  });
};

export function isChildrenSelected(children, keys: string[]) {
  let find = false;
  function loop(_children) {
    if (!_children || find) {
      return;
    }
    React.Children.forEach(_children, (c) => {
      if (c && c.props && c.type && !find) {
        const menuType = c.type.menuType;
        const selectable = c.props.selectable;
        if (menuType === 'MenuItem') {
          find = keys.indexOf(c.key) !== -1;
        } else if (menuType === 'SubMenu' && selectable) {
          find = keys.indexOf(c.key) !== -1;
        }
        if (!find && c.props.children) {
          loop(c.props.children);
        }
      }
    });
  }
  loop(children);
  return find;
}
