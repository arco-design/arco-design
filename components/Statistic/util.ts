import { padStart } from '../_util/pad';

const units: [string, number][] = [
  ['Y', 1000 * 60 * 60 * 24 * 365], // years
  ['M', 1000 * 60 * 60 * 24 * 30], // months
  ['D', 1000 * 60 * 60 * 24], // days
  ['H', 1000 * 60 * 60], // hours
  ['m', 1000 * 60], // minutes
  ['s', 1000], // seconds
  ['S', 1], // million seconds
];

export function getDateString(millisecond: number, format: string) {
  let leftMillisecond: number = millisecond;
  return units.reduce((current, [name, unit]) => {
    if (current.indexOf(name) !== -1) {
      const value = Math.floor(leftMillisecond / unit);
      leftMillisecond -= value * unit;
      return current.replace(new RegExp(`${name}+`, 'g'), (match: string) => {
        const len = match.length;
        return padStart(value.toString(), len, '0');
      });
    }
    return current;
  }, format);
}
