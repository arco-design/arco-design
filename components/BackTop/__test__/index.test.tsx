import React from 'react';
import { fireEvent, render, sleep } from '../../../tests/util';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import BackTop from '..';

mountTest(BackTop);
componentConfigTest(BackTop, 'BackTop');

describe('BackTop', () => {
  it('click button scroll to top', async () => {
    const onClick = jest.fn();

    const window = {
      scrollTop: 400,
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
    };

    const component = render(<BackTop onClick={onClick} target={() => window as any} />);
    fireEvent.click(component.find('.arco-backtop')[0]);
    await sleep(500);
    expect(window.scrollTop).toBe(0);
    expect(onClick).toHaveBeenCalled();
    component.unmount();
    expect(window.removeEventListener).toHaveBeenCalled();
  });
});
