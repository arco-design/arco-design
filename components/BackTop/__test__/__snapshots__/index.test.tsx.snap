// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.BackTop default 1`] = `
<div
  class="arco-backtop"
/>
`;

exports[`ConfigProvider componentConfig.BackTop set className globally 1`] = `
<div
  class="arco-backtop global-default-class"
/>
`;

exports[`ConfigProvider componentConfig.BackTop set className globally and component set className itself. 1`] = `
<div
  class="arco-backtop component-class"
/>
`;

exports[`ConfigProvider componentConfig.BackTop set data-* property 1`] = `
<div
  class="arco-backtop"
  data-test-name="BackTop"
/>
`;
